### DISCUSSION TITLE
静态库程序构建报错，但动态库程序没问题

### DISCUSSION THREAD

**Discussion by AomaYple** [xmake-io/xmake#3162] [Q&A] [answered]
### Xmake 版本

2.7.3

### 操作系统版本和架构

windows11 22h2

### 描述问题

静态库程序构建报错，但动态库程序没问题
![1](https://user-images.githubusercontent.com/102885957/206881587-a3e0ef5c-aaa1-4461-b49c-5eab1de8b2d4.png)
![2](https://user-images.githubusercontent.com/102885957/206881589-0da15990-b5a4-41e6-8337-2f5e1d905c6e.png)

### 期待的结果

静态库程序构建应该成功

### 工程配置

[untitled.zip](https://github.com/xmake-io/xmake/files/10201279/untitled.zip)

**Comment by waruqi** [comment]
mingw clang最近才支持，更新到 dev

xmake update dev

https://github.com/xmake-io/xmake/issues/3060

然后删了自定义的 toolchain，用内置的 clang toolchain

**Comment by AomaYple** [comment]
> mingw clang最近才支持，更新到 dev
> 
> xmake update dev
> 
> #3060
> 
> 然后删了自定义的 toolchain，用内置的 clang toolchain

使用内置工具链的lua语法是什么呀

**Comment by waruqi** [comment]
set_toolchains("clang") 或者 xmake f --toolchain=clang

**Comment by AomaYple** [comment]
> set_toolchains("clang") 或者 xmake f --toolchain=clang

![image](https://user-images.githubusercontent.com/102885957/206891538-011d90f9-065e-414b-88ea-2c4be14a7a5a.png)
卡在这不动了，不会我网速太慢了吧

**Comment by waruqi** [comment]
自己命令行执行下看看

**Comment by AomaYple** [comment]
> 自己命令行执行下看看

![image](https://user-images.githubusercontent.com/102885957/206891812-a2e8086d-f7e8-4ff8-b851-85fa8a883200.png)
也是一直卡这里，xmake版本会影响这个吗

**Comment by waruqi** [comment]
参考这个操作下，给我个栈 

https://github.com/xmake-io/xmake/issues/2033

**Comment by AomaYple** [comment]
> 参考这个操作下，给我个栈
> 
> #2033

是在xmake.lua所在目录下创建一个test.lua文件并且文件内容是
function main()
    io.read()
end
这个吗。
命令行怎么打

**Comment by waruqi** [comment]
不是，参考 https://github.com/xmake-io/xmake/issues/2033 

里面有详细说明

**Comment by waruqi** [comment]
XMAKE_PROFILE=stuck xmake f --toolchain=clang -vD -c
然后卡住后 ctrl-c

**Comment by AomaYple** [comment]
> XMAKE_PROFILE=stuck xmake f --toolchain=clang -vD -c 然后卡住后 ctrl-c

![image](https://user-images.githubusercontent.com/102885957/206892939-277595a7-5bfe-491d-9e45-4d5ff549962a.png)

**Comment by waruqi** [comment]
不要照搬 win下环境变量用 set 设置

**Comment by waruqi** [comment]
set XMAKE_PROFILE=stuck
xmake f --toolchain=clang -vD -c

**Comment by AomaYple** [comment]
> set XMAKE_PROFILE=stuck xmake f --toolchain=clang -vD -c
![image](https://user-images.githubusercontent.com/102885957/206893256-b445728c-285c-42ff-b416-a915bf3ba45e.png)
这样？

**Comment by waruqi** [comment]
锁住了，你有多个 xmake 进程，在同时操作编译这个项目

退掉所有 xmake进程，clion 什么的全退干净后再试

**Comment by AomaYple** [comment]
![image](https://user-images.githubusercontent.com/102885957/206893528-3b07f482-bca9-4caa-bda0-39dd1e0058e0.png)

**Comment by AomaYple** [comment]
![image](https://user-images.githubusercontent.com/102885957/206893568-3d02dc30-0c84-467a-9ea5-25672697c966.png)

**Comment by waruqi** [comment]
你到底是要用 vs 还是 clang

你 vscode 里面，我看用的 msys2 下的 clang ，为啥命令行下不同样到 msys/clang 下跑？

那就同样进到 msys 的 clang 环境执行，默认 cmd 当然没 clang 环境。。而且你也没装 vs

**Comment by waruqi** [comment]
先确认你要用什么工具链，vs, msys2/mingw, msys2/clang 不同工具链，都是不一样的

**Comment by AomaYple** [comment]
> 先确认你要用什么工具链，vs, msys2/mingw, msys2/clang 不同工具链，都是不一样的

不是不用自定义工具链，用xmake内置工具链吗，我的ide是clion，这不是使用内置工具链clang吗

**Comment by waruqi** [comment]
我是说你现在安装了什么工具链？mingw? clang? 不同的工具链，不同的配置方式，我先要确认你装了啥，打算用哪个？你装的工具链不同，不同的用法

我又不知道你想用哪个，只能根据你最早配置了 d:/msys2/clang64 

猜测你想用 msys2 下的 clang 环境，那你就得进到 msys2 shell 环境执行 xmake 命令，而不是 cmd

**Comment by AomaYple** [comment]
> 

成功了！进msys2 shell使用xmake构建静态库成功了

**Comment by AomaYple** [comment]
好像msys2的clang只能进clang64环境使用xmake才能构建成功

**Comment by AomaYple** [comment]
> 我是说你现在安装了什么工具链？mingw? clang? 不同的工具链，不同的配置方式，我先要确认你装了啥，打算用哪个？你装的工具链不同，不同的用法
> 
> 我又不知道你想用哪个，只能根据你最早配置了 d:/msys2/clang64
> 
> 猜测你想用 msys2 下的 clang 环境，那你就得进到 msys2 shell 环境执行 xmake 命令，而不是 cmd

大佬内置工具链是指xmake自带的工具链吗，如果是怎么使用

**Comment by waruqi** [comment]
内置工具链指的是工具链配置，xmake 本身不会自带工具链环境，你装了什么，就用什么，不用自己定义toolchain

想要自动拉取集成工具链，走包模式，参考

https://github.com/xmake-io/xmake/blob/master/tests/projects/package/toolchain_llvm/xmake.lua

**Reply by AomaYple** [reply]
> 内置工具链指的是工具链配置，xmake 本身不会自带工具链环境，你装了什么，就用什么，不用自己定义toolchain
> 
> 想要自动拉取集成工具链，走包模式，参考
> 
> https://github.com/xmake-io/xmake/blob/master/tests/projects/package/toolchain_llvm/xmake.lua

![image](https://user-images.githubusercontent.com/102885957/206901890-a757eb99-f9c5-433f-8096-c144dea6e96a.png)
是网不好吗

**Reply by waruqi** [reply]
上面提示很明显 ssl 证书问题，curl的问题，跟 xmake无关

**Reply by waruqi** [reply]
看下 https://xmake.io/#/zh-cn/guide/faq?id=%e4%b8%8b%e8%bd%bd%e5%8c%85%e6%8f%90%e7%a4%ba%e8%af%81%e4%b9%a6%e6%a0%a1%e9%aa%8c%e5%a4%b1%e8%b4%a5%e6%80%8e%e4%b9%88%e5%8a%9e%ef%bc%9f

**Comment by AomaYple** [comment, marked as answer]
总算懂了，如果要使用msys2的工具链，xmake也要用msys2的才行
